### PR TITLE
fix: crashes caused by invalid saved layout JSON and show toasts for recovery

### PR DESCRIPTION
**Pull request by tonypzy**
## Description

This PR adds defensive JSON parsing (`safeJsonParse`) to handle malformed JSON and prevent frontend runtime exceptions. 

## Motivation and Context

The following malformed JSON cases are now safely handled by `safeJsonParse`:

```
localStorage.setItem('envIDs', '["main", "demo"'); # Truncated JSON
localStorage.setItem('layout_main_pane_1', 'undefined'); # Non-JSON text
localStorage.setItem('layout_main_pane_1', '{x: 0, y: 0}'); # Invalid object syntax
localStorage.setItem('envIDs', ''); # Empty persisted value
```

## How Has This Been Tested?

Check code manually

## Screenshots (if appropriate):
N/A

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tonypzy** [comment]
@vedansh-5 @rajnisht7 @Manik-Khajuria-5 @Jayantparashar10 please review

**Comment by Debajeet-1411** [comment]
@tonypzy Nice improvement. Consider adding a Cypress test for malformed localStorage JSON to verify the recovery path and guard against future regressions.

**Review by Debajeet-1411 — COMMENTED** [review]
One potential edge case came to mind—I've left an inline comment with the details.

File: js/main.js [hunk 1]
```diff
@@ -55,7 +66,10 @@ if (ACTIVE_ENV !== '') {
     use_envs = [ACTIVE_ENV];
   }
 } else {
-  use_envs = JSON.parse(localStorage.getItem('envIDs')) || ['main'];
+  let storedEnvIDs = safeJsonParse(localStorage.getItem('envIDs'), null, () => {
```
Comment (Debajeet-1411) [inline js/main.js:73]: If envIDs contains valid JSON but not an array (e.g. "{}"), safeJsonParse succeeds, so the invalid value isn't removed from localStorage. The code falls back to ['main'], but the stale entry remains and will trigger the same fallback on every load. It might be worth clearing the stored value whenever the parsed value fails validation.

**Review by Jayantparashar10 — APPROVED** [review]

File: js/main.js (hunk 1 above)
Comment (tonypzy) [reply js/main.js:73]: Thanks for the careful review. This PR is scoped to detect malformed JSON values. We can create a new PR to handle the case you mentioned, where the value is valid but semantically incorrect.

**Comment by tonypzy** [comment]
> @tonypzy Nice improvement. Consider adding a Cypress test for malformed localStorage JSON to verify the recovery path and guard against future regressions.

cytest is not meant for handling such trivial cases. it is primarily designed to verify higher-level functionality, such as plotting features and screenshot pixel comparisons.

**Comment by tonypzy** [comment]
> @tonypzy Nice improvement. Consider adding a Cypress test for malformed localStorage JSON to verify the recovery path and guard against future regressions.

Or we can add to pytest #1543

**Review by Manik-Khajuria-5 — APPROVED** [review]

**Comment by Debajeet-1411** [comment]
@tonypzy Thanks for the clarification. That makes sense-this is better suited as a unit test than an end-to-end Cypress test.

**Review by rajnisht7 — COMMENTED** [review]
@tonypzy it works good in ui but what about adding info for the same

**Review by Manik-Khajuria-5 — APPROVED** [review]
@tonypzy I agree with @rajnisht7 idea of adding in the ui section also

**Comment by tonypzy** [comment]
> @tonypzy it works good in ui but what about adding info for the same

This is the toast effect.
image

**Review by Saksham-Sirohi — COMMENTED** [review]
Copilot flagged the following:

- No blocking issues.

**Review by rajnisht7 — APPROVED** [review]

**Review by vedansh-5 — APPROVED** [review]
